src/Error/ErrorTrap.php

```diff
@@ -0,0 +1,254 @@
+    ): bool {
+        if (!(error_reporting() & $code)) {
+            return false;
+        }
+        $trace = Debugger::trace(['start' => 1, 'format' => 'points']);
+        $error = new PhpError($code, $description, $file, $line, $trace);
+
+        $renderer = $this->renderer();
+        $logger = $this->logger();
+
+        try {
+            // Log first incase rendering or callbacks fail.
+            $logger->logMessage($error->getLabel(), $error->getMessage());
+
+            foreach ($this->callbacks as $callback) {
+                $callback($error);
+            }
+            $renderer->write($renderer->render($error));
+        } catch (Exception $e) {
+            $logger->logMessage('error', 'Could not render error. Got: ' . $e->getMessage());
+
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Get an instance of the renderer.
+     *
+     * @return \Cake\Error\ErrorRendererInterface
+     */
+    public function renderer(): ErrorRendererInterface
+    {
+        $class = $this->getConfig('errorRenderer');
+
+        /** @var \Cake\Error\ErrorRendererInterface $instance */
+        $instance = new $class($this->_config);
+
+        return $instance;
+    }
+
+    /**
+     * Get an instance of the logger.
+     *
+     * @return \Cake\Error\ErrorLoggerInterface
+     */
+    public function logger(): ErrorLoggerInterface
+    {
+        $class = $this->getConfig('logger');
+
+        /** @var \Cake\Error\ErrorLoggerInterface $instance */
+        $instance = new $class($this->_config);
+
+        return $instance;
+    }
+
+    /**
+     * Change the error renderer
+     *
+     * @param class-string<\Cake\Error\ErrorRendererInterface> $class The class to use as a renderer.
+     * @return $this
+     */
+    public function setErrorRenderer(string $class)
+    {
+        if (!in_array(ErrorRendererInterface::class, class_implements($class))) {
+            throw new InvalidArgumentException(
+                "Cannot use {$class} as an error renderer. It must implement \Cake\Error\ErrorRendererInterface."
+            );
+        }
+        $this->setConfig('errorRenderer', $class);
+
+        return $this;
+    }
+
+    /**
+     * Set the PHP error reporting level
+     *
+     * @param int $level The PHP error reporting value to use.
+     * @return $this
+     */
+    public function setLevel(int $level)
+    {
+        if ($this->registered) {
+            throw new LogicException('Cannot change level after an ErrorTrap has been registered.');
+        }
+        $this->setConfig('level', $level);
+
+        return $this;
+    }
+
+    /**
+     * Set the Error logging implementation
+     *
+     * When the logger is constructed it will be passed
+     * the current options array.
+     *
+     * @param class-string<\Cake\Error\ErrorLoggerInterface> $class The logging class to use.
+     * @return $this
+     */
+    public function setLogger(string $class)
```

Comment:
`setLoggerClass()` would be a more intuitive name. At other places in the framework `setLogger()` is used to set an instance.